includes/embeds/class-amp-gallery-embed.php

```diff
@@ -241,47 +230,38 @@ public function render( $args ) {
 				$image_atts['srcset'] = $props['srcset'];
 			}
 
-			$this_aspect_ratio = $props['width'] / $props['height'];
-			if ( $this_aspect_ratio > $max_aspect_ratio ) {
-				$max_aspect_ratio = $this_aspect_ratio;
-				$carousel_width   = $props['width'];
-				$carousel_height  = $props['height'];
-			}
-
 			if ( ! empty( $args['lightbox'] ) ) {
 				$image_atts['lightbox'] = '';
-				$image_atts['on']       = 'tap:' . AMP_Img_Sanitizer::AMP_IMAGE_LIGHTBOX_ID;
-				$image_atts['role']     = 'button';
-				$image_atts['tabindex'] = 0;
 			}
-			$image = AMP_HTML_Utils::build_tag(
+			$image = AMP_DOM_Utils::create_node(
+				$dom,
 				'amp-img',
 				$image_atts
 			);
 
 			if ( ! empty( $props['href'] ) ) {
-				$image = AMP_HTML_Utils::build_tag(
+				$previous_image = $image;
+				$image          = AMP_DOM_Utils::create_node(
+					$dom,
 					'a',
 					[
 						'href' => $props['href'],
-					],
-					$image
+					]
 				);
+				$image->appendChild( $previous_image );
 			}
 
-			$images[] = $image;
+			$caption = isset( $props['id'] ) ? wp_get_attachment_caption( $props['id'] ) : '';
+			$images->add( $image, $caption );
 		}
 
-		return AMP_HTML_Utils::build_tag(
-			'amp-carousel',
-			[
-				'width'  => $carousel_width,
-				'height' => $carousel_height,
-				'type'   => 'slides',
-				'layout' => 'responsive',
-			],
-			implode( PHP_EOL, $images )
-		);
+		$amp_carousel  = new Carousel( $dom );
```

Comment:
The `Carousel` does not behave like an object, but rather like a collection of procedural functions:
- create $carousel
- use $carousel on $image_list to create node

Note that the carousel on itself is not functional, and it can be used with multiple image lists. It is a procedural function collection instead of an object.

A `Carousel` does not really make sense without an image list (even if only an empty one), so a better OOP structure would be to accept the image list in the constructor.

Then, the "API" would rather look like this:

```php
$amp_carousel = new Carousel( $dom, $image_list );
$dom->appendChild( $amp_carousel->as_dom_element() );
```

If you want to get fancy, you could even let `AMP\Carousel` extend `DOMElement`, so you could just simply use `$dom->appendChild( $amp_carousel );`.

Reply:
Thanks, https://github.com/ampproject/amp-wp/pull/3659/commits/2b4a98fd6ae549430da57883aea4d60c0bf66945 changes the constructor to accept the `$dom` and `$image_list`. Then, it can get the `DOMElement` in a separate function.

Reply:
Is that what you had in mind?